### PR TITLE
$upstream support for container registry address

### PR DESCRIPTION
**Pull request by huguesBouvier**
$upstream support for container address.
1. Support in azure portal
2. Support in aziot-edged

Test:
Created a registry on lvl5 with a password then put $upstream in config and $upstream in portal. Checked that image got pulled.

Put a wrong password in portal, checked images failed with unauthorized (to confirm image were getting pulled from the right place)
Put a wrong password in config, checked edgeAgent image failed to e pulled with unauthorized (to confirm image were getting pulled from the right place)

Removed $upstream and tried to pull from edgeBuilds directly to make sure nothing was broken by that PR.

### CODE REVIEW

File: edgelet/edgelet-core/src/module.rs [hunk 1]
```diff
@@ -181,6 +181,7 @@ where
 /// time, but for now, this will have to do...
 pub trait NestedEdgeBodge {
     fn parent_hostname_resolve_image(&mut self, parent_hostname: &str);
+    fn parent_hostname_resolve_auth(&mut self, parent_hostname: &str);
```
Comment (arsing) [inline edgelet/edgelet-core/src/module.rs:184]: Just rename the existing method to `parent_hostname_resolve` and have it fixup both the image and the auth.

**Review by arsing — CHANGES_REQUESTED** [review]
As you said, `check_agent_image.rs` is already doing the substitution. But you should still change it to use this new method instead of doing the substitution manually.

Though if you merge the two methods like I suggested in my review comment, then you'd delete the substitution code from that file instead.

File: edgelet/edgelet-core/src/module.rs (hunk 1 above)
Comment (huguesBouvier) [reply edgelet/edgelet-core/src/module.rs:184]: Done!

**Comment by huguesBouvier** [comment]
> As you said, `check_agent_image.rs` is already doing the substitution. But you should still change it to use this new method instead of doing the substitution manually.
> 
> Though if you merge the two methods like I suggested in my review comment, then you'd delete the substitution code from that file instead.

That's good point, thanks!

File: edgelet/edgelet-docker/src/config.rs [hunk 2]
```diff
@@ -94,10 +94,37 @@ impl DockerConfig {
...
+            new_auth.set_username(username.to_string());
+        }
+
+        if let Some(email) = auth.email() {
+            new_auth.set_email(email.to_string());
+        }
+
+        if let Some(password) = auth.password() {
+            new_auth.set_password(password.to_string());
+        }
```
Comment (arsing) [inline edgelet/edgelet-docker/src/config.rs:119]: What's the point of doing all this outside of line 122-124 ?

Comment (huguesBouvier) [reply edgelet/edgelet-docker/src/config.rs]: Good catch, sorry about that, stupid mistake.

File: edgelet/edgelet-docker/src/config.rs [hunk 3]
```diff
@@ -94,10 +94,37 @@ impl DockerConfig {
...
+                    new_auth.set_username(username.to_string());
+                }
+        
+                if let Some(email) = auth.email() {
+                    new_auth.set_email(email.to_string());
+                }
+        
+                if let Some(password) = auth.password() {
+                    new_auth.set_password(password.to_string());
+                }
```
Comment (arsing) [inline edgelet/edgelet-docker/src/config.rs:121]: And do you need to do any of this instead of just `self.auth.set_serveraddress(format!(...));` ?

Comment (huguesBouvier) [reply edgelet/edgelet-docker/src/config.rs]: Fixed it, thanks for the help

File: edgelet/aziot-edged/src/lib.rs [hunk 4]
```diff
@@ -277,7 +277,7 @@ where
                             .unwrap_or(TRUST_BUNDLE_ALIAS)
                             .to_string(),
                         settings
-                            .manifest_trust_bundle_cert()
+                            .trust_bundle_cert()
```
Comment (arsing) [inline edgelet/aziot-edged/src/lib.rs:280]: Is this correct?

Comment (huguesBouvier) [reply edgelet/aziot-edged/src/lib.rs:280]: No it is not, I am not sure how that got in. It was not in previous PR: https://github.com/Azure/iotedge/commit/5e93f5834d51c404d82b6ec1e9213b611b725d84

**Review by arsing — APPROVED** [review]

**Pull request merged**